File: cli/reltime.go
package main

import (
	"fmt"
	"time"
)

// relTime formats a duration as a human-friendly relative time string
// (e.g. "2m ago", "3h ago", "5d ago"). Falls back to the input for zero time.
func relTime(t time.Time) string {
	if t.IsZero() {
		return "never"
	}
	d := time.Since(t)
	switch {
	case d < 10*time.Second:
		return "just now"
	case d < time.Minute:
		return fmt.Sprintf("%ds ago", int(d.Seconds()))
	case d < time.Hour:
		return fmt.Sprintf("%dm ago", int(d.Minutes()))
	case d < 24*time.Hour:
		return fmt.Sprintf("%dh ago", int(d.Hours()))
	case d < 7*24*time.Hour:
		return fmt.Sprintf("%dd ago", int(d.Hours()/24))
	case d < 30*24*time.Hour:
		return fmt.Sprintf("%dw ago", int(d.Hours()/24/7))
	default:
		return t.Format("2006-01-02")
	}
}

// parseTimeOrZero parses an ISO/RFC3339 timestamp, returning zero time on failure.
func parseTimeOrZero(s string) time.Time {
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339, "2006-01-02T15:04:05Z"} {
		if t, err := time.Parse(layout, s); err == nil {
			return t
		}
	}
	return time.Time{}
}

File: src/web/src/reltime.ts
/** Format an ISO timestamp as a human-friendly relative time string. */
export function formatRelativeTime(iso: string): string {
  const t = new Date(iso);
  if (isNaN(t.getTime())) return iso; // unparseable, return as-is
  const d = Date.now() - t.getTime();
  const s = Math.floor(d / 1000);
  if (s < 10) return "just now";
  if (s < 60) return `${s}s ago`;
  const m = Math.floor(s / 60);
  if (m < 60) return `${m}m ago`;
  const h = Math.floor(m / 60);
  if (h < 24) return `${h}h ago`;
  const dd = Math.floor(h / 24);
  if (dd < 7) return `${dd}d ago`;
  if (dd < 30) return `${Math.floor(dd / 7)}w ago`;
  return t.toLocaleDateString();
}
